pkg/cmd/cli/cmd/new/sourcelookup.go

```diff
@@ -0,0 +1,185 @@
+package new
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/openshift/origin/pkg/generate/app"
+	"github.com/openshift/origin/pkg/generate/dockerfile"
+	"github.com/openshift/origin/pkg/generate/source"
+)
+
+var (
+	argumentGit         = regexp.MustCompile("^(http://|https://|git@|git://).*\\.git(?:#([a-zA-Z0-9]*))?$")
+	argumentGitProtocol = regexp.MustCompile("^git@")
+	argumentPath        = regexp.MustCompile("^\\.")
+)
+
+func isPossibleSourceRepository(s string) bool {
+	return argumentGit.MatchString(s) || argumentGitProtocol.MatchString(s) || argumentPath.MatchString(s)
+}
+
+// SourceRepository represents an code repository that may be the target of a build.
+type SourceRepository struct {
+	location string
+	url      url.URL
+
+	usedBy          []ComponentReference
+	buildWithDocker bool
+}
+
+// NewSourceRepository creates a reference to a local or remote source code repository from
+// a URL or path.
+func NewSourceRepository(s string) (*SourceRepository, error) {
+	var location *url.URL
+	switch {
+	case strings.HasPrefix(s, "git@"):
+		base := "git://" + strings.TrimPrefix(s, "git@")
+		url, err := url.Parse(base)
+		if err != nil {
+			return nil, err
+		}
+		location = url
+
+	default:
+		uri, err := url.Parse(s)
+		if err != nil {
+			return nil, err
+		}
+
+		if uri.Scheme == "" {
+			path := s
+			ref := ""
+			segments := strings.SplitN(path, "#", 2)
+			if len(segments) == 2 {
+				path, ref = segments[0], segments[1]
+			}
+			path, err := filepath.Abs(path)
+			if err != nil {
+				return nil, err
+			}
+			uri = &url.URL{
+				Scheme:   "file",
+				Path:     path,
+				Fragment: ref,
+			}
+		}
+
+		location = uri
+	}
+	return &SourceRepository{
+		location: s,
+		url:      *location,
+	}, nil
+}
+
+func (r *SourceRepository) UsedBy(ref ComponentReference) {
+	r.usedBy = append(r.usedBy, ref)
+}
+
+func (r *SourceRepository) Remote() bool {
+	return r.url.Scheme != "file"
+}
+
+func (r *SourceRepository) InUse() bool {
+	return len(r.usedBy) > 0
+}
+
+func (r *SourceRepository) BuildWithDocker() {
+	r.buildWithDocker = true
+}
+
+func (r *SourceRepository) IsDockerBuild() bool {
+	return r.buildWithDocker
+}
+
+func (r *SourceRepository) String() string {
+	return r.location
+}
+
+func (r *SourceRepository) LocalPath() (string, error) {
+	switch {
+	case r.url.Scheme == "file":
+		return r.url.Path, nil
+		// TODO: implement other types
+		// TODO: lazy cache (predictably?)
+	default:
+
+		return "", fmt.Errorf("reading local repositories is not implemented: %q", r.location)
```

Comment:
I would think that we do want to read local repositories as long as we can determine a valid remote url